File: famicom-dumper/mappers-lua/vrc6a.lua
--
-- VRC6a mapper
-- http://wiki.nesdev.com/w/index.php/VRC6
--  Akumajou Densetsu
--

MapperName = "VRC6a"
MapperNumber = 24
DefaultPrgSize = 256 * 1024
DefaultChrSize = 256 * 1024

function DumpPrg(size)
	local banks = size / 0x4000
	for b = 0, banks-1 do
		print("Reading PRG banks #" .. tostring(b) .. "...")
		WriteCpu(0x8000, {b})
		ReadAddPrg(0x8000, 0x4000)
	end
end

function DumpChr(size)
    WriteCpu(0xB003,{0x00})
	local banks = size / 0x0400
	for b = 0, banks-1, 8 do
		print("Reading CHR banks #" .. tostring(b) .. ", #" .. tostring(b+1) .. ", #" .. tostring(b+2) .. ", #" .. tostring(b+3) .. ", #" .. tostring(b+4) .. ", #" .. tostring(b+5) .. ", #" .. tostring(b+6) .. " and #" .. tostring(b+7) .. "...")
		WriteCpu(0xD000, {b, b+1, b+2, b+3})
		WriteCpu(0xE000, {b+4, b+5, b+6, b+7})
		ReadAddChr(0x0000, 0x2000)
	end
end

function EnablePrgRam(size)
	WriteCpu(0xB003, {0x80})
end

File: FamicomDumper/mappers/VRC6a.cs
class VRC6a : IMapper
{
    public string Name { get => "VRC6a"; }
    public int Number { get => 24; }
    public int DefaultPrgSize { get => 256 * 1024; }
    public int DefaultChrSize { get => 256 * 1024; }

    public void DumpPrg(IFamicomDumperConnection dumper, List<byte> data, int size)
    {
        var banks = size / 0x4000;

        for (var bank = 0; bank < banks; bank++)
        {
            Console.Write($"Reading PRG bank #{bank}/{banks}... ");
            dumper.WriteCpu(0x8000, (byte)bank);
            data.AddRange(dumper.ReadCpu(0x8000, 0x4000));
            Console.WriteLine("OK");
        }
    }

    public void DumpChr(IFamicomDumperConnection dumper, List<byte> data, int size)
    {
        dumper.WriteCpu(0xB003, 0xE0);
        var banks = size / 0x400;

        for (var bank = 0; bank < banks; bank++)
        {
            Console.Write($"Reading CHR bank #{bank}/{banks}... ");
            dumper.WriteCpu(0xD000, (byte)bank);
            data.AddRange(dumper.ReadPpu(0x0000, 0x0400));
            Console.WriteLine("OK");
        }
    }

    public void EnablePrgRam(IFamicomDumperConnection dumper)
    {
        dumper.WriteCpu(0xB003, 0xE0);
    }

    public MirroringType GetMirroring(IFamicomDumperConnection dumper)
    {
        return MirroringType.MapperControlled;
    }
}
